docs/components-and-props.md (#14)
reactjs/de.reactjs.org: content/docs/components-and-props.md

Du kannst auch [ES6 Klassen](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) benutzen um Komponenten zu definieren:

```js
class Welcome extends React.Component {
  render() {
    return <h1>Hallo {this.props.name}</h1>;
  }
}
```

Die beiden obrigen Komponenten sind aus der Sicht von React identisch.

Klassen haben noch ein paar zusätzliche Eigenschaften, diese besprechen wir aber erst im [nächsten Abschnitt](/docs/state-and-lifecycle.html). Bis dahin werden wir für die Funktionskomponenten wegen ihrer Übersichtlichkeit nehmen.

## Eine Komponente rendern {#rendering-a-component}

Bis jetzt haben wir nur React-Elemente kennegelernt, dir DOM-Tags repräsentiert haben:

```js
const element = <div />;
```

Elemente können aber auch benutzerdefinierte Komponenten darstellen:

```js
const element = <Welcome name="Sara" />;
```

React übergibt, wenn es ein Element als benutzerdefinierte Komponente erkennt, alle JSX Attribute als ein einziges Objekt. Dies sind die sogenannten "props" (Eigenschaften).

> Zum Beispiel rendert dieser Code "Hallo Sara" auf die Seite:

Author: I was also thinking about it, but I stumpled upon these lines
```
<Welcome name="Sara" />
<Welcome name="Cahal" />
<Welcome name="Edite" />
```
and I left it as it is.

);
```

[Auf CodePen ausprobieren](codepen://components-and-props/rendering-a-component)

Lass uns mal zusammen fassen, was in diesem Beispiel passiert:

1. Wir rufen `ReactDOM.render()` mit dem React-Element `<Welcome name="Sara" />` auf.
2. React ruft die `Welcome` Komponente mit den Eigenschaften (props) `{name: 'Sara'}` auf.
3. Unsere `Welcome` Komponente gibt als Ergebnis `<h1>Hallo Sara</h1>` zurück.
4. React aktualsiert effektiv das DOM um `<h1>Hallo Sara</h1>` abzugleichen.

> >**Hinweis:** Starte Namen von Komponenten immer mit Großbuchstaben.

Reviewer: ```suggestion
>**Hinweis:** Beginne den Namen von Komponenten immer mit einem Großbuchstaben.
```

>

> >React behandelt Komponenten, die mit Kleinbuchstaben beginnen, als DOM-Tags. Zum Beispiel stellt `<div />` ein HTML div-Tag dar, aber `<Welcome />` stellt eine Komponente dar und erfordert, dass `Welcome` im Scope ist.

Reviewer: ```suggestion
>React behandelt Komponenten, die mit Kleinbuchstaben beginnen, als DOM-Tags. Zum Beispiel stellt `<div />` ein HTML div-Tag dar, `<Welcome />` hingegen ist eine Komponente und erfordert, dass `Welcome` im Scope ist.
```

>
>Bitte lese [JSX im Detail](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized), um mehr über diese Konvention zu erfahren.

## Komponenten zusammensetzen {#composing-components}

Komponenten können bei ihrer Ausgabe auf andere Komponenten verweisen.
So können wir für jedes Detaillevel die selben abstrahierten Komponenten wiederverwenden. Denn sowohl Buttons, Formulare als auch Screens, werden in React-Apps allgemein als Komponenten bezeichnet.

> Zum Beispiel können wir die `App` Komponente mehrere Male `Welcome` rendern lassen:

Reviewer: ```suggestion
Zum Beispiel können wir die `App` Komponente mehrmals `Welcome` rendern lassen:
```

);
```

[Auf CodePen ausprobieren](codepen://components-and-props/composing-components)

> Typischerweise haben neue React-Apps eine einzige `App` Komponente an erste Stelle. Wenn du aber React in ein bestehendes Projekt integrierst, fängst du wahrscheinlich von unten nach oben (bottom-up) an und erstellst Komponenten wie `Button`, dabei arbeitest dich Schritt für Schritt die View-Hierarchy nach oben.

Reviewer: #10 should we add View-Hierarchy?

Author: Yes and I think we can write `View-Hierarchie`.

## Komponenten auslagern {#extracting-components}

Hab keine Angst vor dem Aufteilen von Komponenten in kleinere Komponenten.

> Nehmen wir mal als Beispiel diese `Comment` Komponente:

Reviewer: ```suggestion
Nehmen wir mal als Beispiel diese `Comment`-Komponente:
```

}
```

[Auf CodePen ausprobieren](codepen://components-and-props/extracting-components)

> Diese Komponente nimmt `author` (ein object), `text` (ein string), and `date` (ein date) als Props entgegen und beschreibt einen Kommentar auf einer Social Media Webseite.

Reviewer: ```suggestion
Diese Komponente nimmt `author` (ein Objekt), `text` (ein String), und `date` (ein date-Objekt) als Props entgegen und beschreibt einen Kommentar auf einer Social Media Webseite.
```

}
```

> Der `Avatar` muss nicht wissen, dass er in innerhalb `Comment` gerendert wird. Darum geben wir dem prop einen gebräuchlicheren namen als: `author` und nennen es `user`.

Reviewer: ```suggestion
Der `Avatar` muss nicht wissen, dass er in innerhalb von `Comment` gerendert wird. Darum geben wir dem prop einen gebräuchlicheren namen als: `author` und nennen es `user`.
```

}
```

> Dadurch können wir `Comment` noch ein wenig mehr vereinfachen:

Reviewer: ```suggestion
Dadurch können wir `Comment` weiter vereinfachen:
```

Reviewer: (optional / opinionated) ;)

}
```

[Auf CodePen ausprobieren](codepen://components-and-props/extracting-components-continued)

> Komponenten zu extrahieren mag sich wie Routinearbeit anfühlen, aber am Ende zahlt es sich für größere Apps aus eine Palette an wiederverwendebaren Komponenten zu haben. Eine gute Fausregel ist es, dass wenn ein Teil der Benutzeroberfläche, wie `Button`, `Panel` oder `Avatar` mehrere Male verwendet werden oder so komplex sind wie `App`, `FeedStory` oder `Comment`, gute Kandidaten für wiederverwendebare Komponenten sind.

Reviewer: ```suggestion
Komponenten zu extrahieren mag sich wie Routinearbeit anfühlen, aber am Ende zahlt es sich für größere Apps aus, eine Palette an wiederverwendebaren Komponenten zu haben. Eine gute Faustregel ist es, dass wenn ein Teil der Benutzeroberfläche, wie `Button`, `Panel` oder `Avatar` öfters verwendet werden oder ähnlich komplex wie `App`, `FeedStory` oder `Comment` sind, gute Kandidaten für wiederverwendbare Komponenten sind.
```

> ## Props und Read-Only {#props-are-read-only}

Reviewer: ```suggestion
## Props sind Read-Only {#props-are-read-only}
```

> Es ist egal ob eine Komponente [als Funktion oder Klasse](#function-and-class-components) deklarierst, sie darf nie ihre eigenen props verändern. Schauen wir uns mal diese `sum` Funktion an:

Reviewer: ```suggestion
Es ist egal, ob eine Komponente [als Funktion oder Klasse](#function-and-class-components) deklariert wird, sie darf nie ihre eigenen props verändern. Schauen wir uns mal diese `sum` Funktion an:
```
